hermes-api/src/main/java/pl/allegro/tech/hermes/api/ErrorCode.java

```diff
@@ -63,7 +63,8 @@ public enum ErrorCode {
     UNKNOWN_MIGRATION(NOT_FOUND),
     INVALID_QUERY(BAD_REQUEST),
     IMPLEMENTATION_ABSENT(NOT_FOUND),
-    MOVING_SUBSCRIPTION_OFFSETS_VALIDATION_ERROR(BAD_REQUEST);
+    MOVING_SUBSCRIPTION_OFFSETS_VALIDATION_ERROR(BAD_REQUEST),
+    SENDING_TO_KAFKA_TIMEOUT(GATEWAY_TIMEOUT);
```

Comment:
I'm on the fence with this status code. On one hand, I like it because it is more specialized than, for example, 503. However, it suggests that Hermes acts just as a proxy to something (which is kinda true ;)), but our clients don't necessarily think about it that way. So, they might assume that a load balancer between Hermes and them has returned it. Additionally, within the total request duration, we perform some processing (of course not too much, and usually Kafka is the source of publish requests latency), so it's not like we just pass requests to an external dependency. 

Do you think 503 is too ambiguous in our context?

Reply:
> Additionally, within the total request duration, we perform some processing (of course not too much, and usually Kafka is the source of publish requests latency), so it's not like we just pass requests to an external dependency.

This error is right now returned when we are pretty much sure that Kafka was the source of latency:

```java
public boolean setTimeoutSendingToKafka() {
        return state.compareAndSet(SENDING_TO_KAFKA_PRODUCER_QUEUE, TIMEOUT_SENDING_TO_KAFKA) || state.compareAndSet(SENDING_TO_KAFKA, TIMEOUT_SENDING_TO_KAFKA);
    }
```

My initial idea was that if clients are monitoring Hermes responses they would be able to tell the difference between 500 and 504. 500 would mean "server side, unexpected error" while 504 would mean "server side, sort-of expected error"? But maybe you are right that 504 should be used exclusively by proxies (say nginx), I would say that 503 is sufficient